### ISSUE TITLE
Client connection can hang in SSL_do_handshake

### ISSUE THREAD

**Issue by miketsts** [yhirose/cpp-httplib#682]
I'm trying to improve connection handling in my client application, and recently play a lot with connecting/disconnecting my network adapter. Since I do it manually during the application run, I may disconnect the network at different points of the code.
It happened to me a few times that the code hangs up indefinitely in `SSL_do_handshake` function. I'm not a specialist in openSSL but from some Google reading I understand that it can be related to (non)blocking I/O settings.
Unfortunately, I can't reproduce this situation very easy - it happened to me only once or twice. I also work with system libssl.so, so it doesn't have debug info.
Does my description make sense and you can give an advice about possible changes? Or I need to debug a bit more?

Thanks - and sorry for a possibly unclear description :-)

**Comment by yhirose** [comment]
@miketsts, thanks for the feedback. If your problem happens inside the OpenSSL library, I guess we can do little on cpp-httplib side. If you do the same operation with curl or other HTTPS client libraries, what happens?

**Comment by miketsts** [comment]
Hi @yhirose , thank you for the answer.

Let me try to explain the problem again. Since the connection to the server involves several steps, the network problem can come during each of them. For example, if a network is disconnected during DNS lookup or socket creation - then the problem I describe does not happen, the usual connection timeout works very well. But if the network failure occurs exactly during the SSL handshake stage - then it waits forever.
Following the above, it's very hard to reproduce the problem. And therefore I'm not sure how relevant is to attempt to use curl - I can't try it. But I still reproduced it using httplib few times - each time occasionally, when I was not ready to debug :-). All I could see by attaching gdb is that current stack was waiting in `SSL_do_handshake()`.

I read that the problem can be overcomed if the SSL BIO could be set as nonblocking. But as I'm not much a specialist in OpenSSL, I'm not sure how to do it right. From my brief looking at httplib code `set_nonblocking()` function does not deal with SSL BIO, but only with the socket itself.  The [documentation for SSL_connect()](https://www.openssl.org/docs/man1.1.1/man3/SSL_connect.html) has some description. As I understand, before calling `SSL_connect()` we need somehow set BIO to nonblocking and wait with `select()`. But I'm still not sure how to do it right, and how to integrate it into httplib design.

If you could give me some hints - I could try to implement it on my side. If you can think of some tests that could reproduce the problem more easily - I would be grateful as well !

**Comment by yhirose** [comment]
@miketsts, thank you for the detailed info. I don't mind to take your suggestion, but I am not planning to work on it unless it can be reproducible because we can't validate if the change really works and the change is very hard to debug.

Of course, you are very much welcome to send me a pull request including the fix and a reproducible unit test since it seems like you are able to reproduce it once in a while. When it comes to switching to non-blocking mode for the check, httplib.h does already take the similar approach for the connection timeout check. So you can get idea from the code.

https://github.com/yhirose/cpp-httplib/blob/79ce6f1745099803a3a109f59ca40490ebe5ad1b/httplib.h#L1953-L1967

Hope it helps!

**Comment by yhirose** [comment]
@miketsts, please let me know your plan about it. If you don't have time to work on it at this point, I'll close it for now because I don't think it's a critical issue for me. And I'll reopen it when you are ready. Thanks!

**Comment by miketsts** [comment]
@yhirose I'm sorry indeed for the too long time it takes to me - I can work on it only on weekends, and I've skipped one weekend as well :)
I hope I have advanced with the solution and soon can be able to send it for review.
I also saw similar problems in #680 and #698 - but didn't look yet at the solutions. So I can't be sure if they are related or not.

I can reproduce the problem only manually with a debugger - I manually disconnect the network after I successfully create a socket but before calling `SSL connect`. If you have an idea how can I automate this for testing - you are most welcome.

The solution probably will be:
1. Convert the socket to nonblocking
2. Set SSL BIO as nonblocking
3. Call SSL_connect iteratively with select.
After that I'm not decided if return the system to blocking mode or leave it unblocking - as probably all subsequent operations will be done in non-blocking mode as well.

**Comment by yhirose** [comment]
@miketsts, 1., 2. and the first part of 3. sound very good to me. But after SSL_connection succeeds, SSL BIO should be back to blocking, because my code totally depends on **blocking**. Hope your attempt will go well!

**Comment by yhirose** [comment]
@miketsts, I have tried your following statement, but I couldn't reproduce it on my Mac...

> I can reproduce the problem only manually with a debugger - I manually disconnect the network after I successfully create a socket but before calling SSL connect.

1) Build `example/server.cc`, and run the server
2) Build `example/client.cc`, and start the client on lldb, set a break point on SSL_connect (or SSL_do_handlshake)
3) Run the client no lldb, then the breakpoint is hit. Then kill the server, and continue the client.
4) Return with `Error::SSLConnection` instead of crash or not blocked by SSL_connect.

**Comment by miketsts** [comment]
Hi @yhirose 

I've commited PR #728 as a first version of the solution. It should be still improved.

Unfortunately, my change causes failure in most of server tests. After some investigation in Google it seems to me that there is a hidden problem in your server implementation.
The symptom of the problem is that SSL_connect returns SSL_ERROR_SYSCALL error, while the `errno` value is 0 (success). Google search says that this can be related to the server closing the socket immediately after connection. I didn't try to debug the server together with the client, as I'm not familiar with its code.
I deliberately left the debug output in my code, like below:
```
SSL_ERROR_SYSCALL: Success
```
If you have a suggestion how this can be handled - you are most welcome.

As for the problem in topic. As I wrote, I can reproduce the disconnection problem manually with debugger, just as you suggested. I put a break point before calling SSL_Connect and disable my Wi-Fi adapter. This seem working now - but it would be nice to be able to reproduce this in unit tests in automatic way. Maybe we can expose the internals in some way and have two separate API calls from a test? Just an idea.

There might be yet another hidden problem that I also didn't investigate. I have compiled my version of openssl-1.1.1h - and it failed to work with the library. It returned `X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY` error from SSL_Connect call, even without my changes. When I reverted to the default system library (I run Ubuntu 18.04), that problem went away.

Of course, I will be glad for any comments, remarks for code style etc.

**Comment by yhirose** [comment]
@miketsts, thank very much for the further investigation and the pull request. I'll put comments about the implementation there.

> If you have a suggestion how this can be handled - you are most welcome.

Right now, I don't have any clue...

> As for the problem in topic. As I wrote, I can reproduce the disconnection problem manually with debugger, just as you suggested.

Could explain the exact steps more clearly, so that I could reproduce it on your machine? I'll follow them step by step. (But, I like the following 'unit test' idea better than this manual way.

> but it would be nice to be able to reproduce this in unit tests in automatic way. Maybe we can expose the internals in some way and have two separate API calls from a test? Just an idea.

Of course It's fantastic if you could add a unit test for this issue! (Then I don't have to struggle with reproducing it any more!)
I am only against the approach to "have two separate API calls". Instead, I prefer keeping the current APIs as they are, and put some debug code with a C++ define symbol like `CPPHTTPLIB_DEBUG`. Could you work on this as a priority, and put a unit test in another pull request? This is a huge step to fix this issue, because I can really start looking into it.

Thank you so much for your help!

**Comment by miketsts** [comment]
@yhirose I'm happy to tell that I think I've fixed the problem I had and updated the PR.
Now all tests are passing on my Ubuntu.
I have removed the disabled unit test from the PR - I still don't have an idea how to cause the network disconnection in a programmatic way.

Unfortunately, I see that the GitHub pipeline falls on Mac with segmentation violation. I don't have a Mac to test this, unfortunately.

**Comment by miketsts** [comment]
Fixed! :)

**Comment by miketsts** [comment]
@yhirose I've got an idea about testing the SSL connect timeout.
The idea is to create an SSL server which calls `bind_to_any_port()` and `listen_after_bind()`, but goes to sleep between the two calls. Meanwhile a client starts up, socket connection succeeds, and it waits for SSL handshake.

I didn't try it yet - will try to find time.

**Comment by yhirose** [comment]
@miketsts, sounds a great idea!
